src/liballoc/raw_vec.rs

```diff
@@ -19,28 +24,18 @@ mod tests;
 /// involved. This type is excellent for building your own data structures like Vec and VecDeque.
 /// In particular:
 ///
-/// * Produces `Unique::empty()` on zero-sized types.
-/// * Produces `Unique::empty()` on zero-length allocations.
 /// * Catches all overflows in capacity computations (promotes them to "capacity overflow" panics).
 /// * Guards against 32-bit systems allocating more than isize::MAX bytes.
 /// * Guards against overflowing your length.
-/// * Aborts on OOM or calls `handle_alloc_error` as applicable.
-/// * Avoids freeing `Unique::empty()`.
+/// * Calls `handle_alloc_error` for fallible allocations.
 /// * Contains a `ptr::Unique` and thus endows the user with all related benefits.
 ///
 /// This type does not in anyway inspect the memory that it manages. When dropped it *will*
 /// free its memory, but it *won't* try to drop its contents. It is up to the user of `RawVec`
 /// to handle the actual things *stored* inside of a `RawVec`.
 ///
-/// Note that a `RawVec` always forces its capacity to be `usize::MAX` for zero-sized types.
-/// This enables you to use capacity-growing logic catch the overflows in your length
-/// that might occur with zero-sized types.
-///
-/// The above means that you need to be careful when round-tripping this type with a
-/// `Box<[T]>`, since `capacity()` won't yield the length. However, `with_capacity`,
-/// `shrink_to_fit`, and `from_box` will actually set `RawVec`'s private capacity
-/// field. This allows zero-sized types to not be special-cased by consumers of
-/// this type.
```

Comment:
I am hesistant to change the guarantees provided by the `RawVec` type. I would prefer if we only changed the internals without affecting the API guarantees.

Reply:
Actually this still holds, the internal capacity field are set with `with_capacity`, `shrink_to_fit`, and `from_box`. I'll re-add this.

Reply:
```rust
/// * Uses the excess returned from the allocator to use the largest available capacity.
/// ...
/// Note that growing a `RawVec` with a zero-sized types always results in an infinite excess
/// represented as `usize::MAX`, so `capacity()` always returns this. This means that you need to be 
/// careful when round-tripping this type with a `Box<[T]>`, since `capacity()` won't yield the 
/// length. However, `with_capacity`, `shrink_to_fit`, and `from_box` will actually set `RawVec`'s 
/// private capacity field. This allows zero-sized types to not be special-cased by consumers of
/// this type.
```

Reply:
Additionally, `into_box` should return `Box<[MaybeUninit<T>]>` instead of `Box<T>`.